### PR TITLE
DAOS-6077 swim: fault injection into SWIM

### PR DESCRIPTION
**Pull request by dmiter**
Add fault injection into SWIM to test the time of failure detection.
To enable this functionality you should set the following YAML file
in D_FI_CONFIG environment.
```
fault_config:
    # SWIM messages drop fault injection
    # probability_x should be 1 always to drop SWIM messages
    # Change probability_y to non-zero to introduce random failures.
    # probability_y: 20  - will introduce 5% of message loss
    # probability_y: 100 - will introduce 1% of message loss
  - id:            65025
    interval:      0
    max_faults:    0
    probability_x: 1
    probability_y: 0
    err_code:      0
    # Comma separated list of SWIM messages drop fault injection arguments:
    # rank=<num> - Set it to the rank number which should be evicted by
    # simulation of network failures.
    # delay=<num> - Set it to the number of seconds on which fault injection
    # will be delayed. It can be useful to complete an initialization and
    # creation of pool/container need for testing.
    argument:     "rank=1,delay=10"
```

### CODE REVIEW

**Comment by berserk-fury** [comment]
Can we include the example on the cover letter of this patch into the [fault-inject-cart.yaml](https://github.com/daos-stack/daos/blob/master/src/cart/utils/fault-inject-cart.yaml) file?

Could you update the [README.env](https://github.com/daos-stack/daos/blob/master/src/cart/README.env) list of the environment variables used in CaRT? That list is missing the definitions of CRT_SWIM_FAIL_ID and CRT_SWIM_FAIL_DELAY.

**Comment by dmiter** [comment]
Hmm,
 ln: failed to create symbolic link ‘/var/lib/jenkins/jenkins-1/docker_1/workspace/daos-stack_daos_PR-3890@3/build/dev/gcc/src/control/src/github.com/daos-stack/daos/src/control’: File exists
re-run helps...

**Comment by dmiter** [comment]
Functional_Hardware_Medium - ERROR: Detected one or more tests that failed archiving! - DAOS-5975

File: src/cart/crt_swim.c
Comment (ashleypittman) [inline src/cart/crt_swim.c:94]: Isn't there an "attribute" field in the yaml file which is intended for parameters like this.

Comment (dmiter) [reply src/cart/crt_swim.c:94]: Do you mean "argument"? Hmm. I can use it but two arguments will complicate parsing...

Comment (ashleypittman) [reply src/cart/crt_swim.c:94]: That would be it.  You could always add a "argument2" to the field for the other one you need.

File: src/cart/crt_swim.c
Comment (ashleypittman) [inline src/cart/crt_swim.c:42]: This should probably call D_SHOULD_FAIL() first, then do the other checks.  Is "id" a rank, it's not clear from the names.

D_SHOUL_FAIL is defined to be efficient if not used, and to do logging if it is so it's useful to call it first.

File: src/cart/crt_swim.c
Comment (ashleypittman) [inline src/cart/crt_swim.c:220]: Probably a warning, and you don't need the "*** " prefix.

File: src/cart/crt_swim.c
Comment (ashleypittman) [inline src/cart/crt_swim.c:327]: fa_err_code comes from the config file here doesn't it?  Why are you taking it's negative value in which case?

File: src/cart/crt_swim.c
Comment (ashleypittman) [inline src/cart/crt_swim.c:332]: And are you selectively freeing something based on the contents of the config file?  Either it needs to be freed or it doesn't.

File: src/cart/crt_swim.c
Comment (ashleypittman) [inline src/cart/crt_swim.c:657]: This doesn't need to be D_ERROR

File: src/cart/crt_swim.c
Comment (ashleypittman) [inline src/cart/crt_swim.c:666]: What if it is NULL? Would that be an error because the fault injection file isn't complete?

**Review by ashleypittman — COMMENTED** [review]
Is there any test code which uses this?

File: src/cart/crt_swim.c
Comment (dmiter) [reply src/cart/crt_swim.c:42]: yes, SWIM id is rank in current implementation.

Comment (dmiter) [reply src/cart/crt_swim.c:42]: Function call is more expensive and I'd like to make it last.

File: src/cart/crt_swim.c
Comment (dmiter) [reply src/cart/crt_swim.c:220]: I definitely would like to see this message even if user specify ERR debug level. I put '***' before all messages related to fault injection. So, you can easily find them out.

File: src/cart/crt_swim.c
Comment (dmiter) [reply src/cart/crt_swim.c:327]: yes, this come from YAML config. fa_err_code is unit32_t. I have to return a negative error.

File: src/cart/crt_swim.c
Comment (dmiter) [reply src/cart/crt_swim.c:332]: This is required because of 'upds' should be free in completion callback or in case of error in upper layer. If we simulate success (rc != 0) we should simulate free like completion callback.

File: src/cart/crt_swim.c
Comment (dmiter) [reply src/cart/crt_swim.c:657]: I definitely would like to see this message even if user specify ERR debug level. This will highlight issues we will have. I should understand why SWIM issue happens.

File: src/cart/crt_swim.c
Comment (dmiter) [reply src/cart/crt_swim.c:666]: If argument is NULL we will use default values which are set in lines 649-652.

File: src/cart/crt_swim.c
Comment (ashleypittman) [reply src/cart/crt_swim.c:42]: D_SHOULD_FAIL() is not a function call, as I said it's deliberately written to be very low-cost when FI is disabled.

File: src/cart/crt_swim.c
Comment (ashleypittman) [reply src/cart/crt_swim.c:327]: That should be changed to int32_t or possibly just int then, the purpose of fa_erro_code is to put daos error numbers in.

File: src/cart/crt_swim.c
Comment (ashleypittman) [reply src/cart/crt_swim.c:332]: That makes sense,  does that mean it also wants to be freed on other error paths, perhaps on L338 for example?

What would it mean if rc was set to zero here?  Would that mean the message is dropped, but anything else it gets replied to with the provided error code?

File: src/cart/crt_swim.c
Comment (ashleypittman) [reply src/cart/crt_swim.c:657]: This is only emitted if FI is enabled?  There is the severity and facility to search for so it certainly doesn't need the *** part.

File: src/cart/crt_swim.c
Comment (dmiter) [reply src/cart/crt_swim.c:42]: Actually my macro join D_SHOULD_FAIL with my addition check. 
D_SHOULD_FAIL() is just line 44 ->  "d_fault_inject && d_should_fail(fa)"

File: src/cart/crt_swim.c
Comment (dmiter) [reply src/cart/crt_swim.c:327]: It will be not consistent anyway. We have definition of DER_* as positive codes. Those error codes returned as "-DER_*".

File: src/cart/utils/fault-inject-cart.yaml
Comment (berserk-fury) [inline src/cart/utils/fault-inject-cart.yaml:58]: I don’t know if these suggestions makes sense.

Is there any chance to support more than one rank for drop injection?
I tried using `"rank=1,rank=2,delay=300"` and it seems that it takes only the later rank value. I mean only rank 2 is affected

`*** SWIM id=2 should fail`

May we have a “duration” parameter? So, we can create transient failures.

File: src/cart/crt_swim.c
Comment (dmiter) [reply src/cart/crt_swim.c:332]: As I mentioned before in case of error caller will free this. In case of error thare is no callback will be called therefore caller should free all data passed to this function.

Comment (dmiter) [reply src/cart/crt_swim.c:332]: rc == 0 for this function means SWIM message was successfully sent and RPC completion callback will be called to free all data.

File: src/cart/crt_swim.c
Comment (dmiter) [reply src/cart/crt_swim.c:657]: Yes, this message will appears only when FI is enabled. For regular code path it never appears.

File: src/cart/utils/fault-inject-cart.yaml
Comment (dmiter) [reply src/cart/utils/fault-inject-cart.yaml:58]: Currently only single rank is supported. Next instance with the same name will overwrite previous one.

I can add  “duration” parameter but is it really make sence for testing? It will not be restarted after duration complete.

Comment (ashleypittman) [reply src/cart/utils/fault-inject-cart.yaml:58]: We can already create transient failures can't we?

Comment (dmiter) [reply src/cart/utils/fault-inject-cart.yaml:58]: Actually we already have "max_faults" for this. It will stop injection after this count fails.

File: src/cart/crt_swim.c
Comment (ashleypittman) [reply src/cart/crt_swim.c:327]: But once they're variables they're always negative, i'd assume you'd put negative values in the config file as that's how they all arrear in log files/messages.

Comment (dmiter) [reply src/cart/crt_swim.c:327]: It's readed by strtoul() now. So, if you specify it as negative in config it will be readed as zero.

File: src/cart/utils/fault-inject-cart.yaml
Comment (berserk-fury) [reply src/cart/utils/fault-inject-cart.yaml:58]: I was trying to understand if we have all the required parameters to reproduce the experiments on the paper pointed out by Liang. It seems that we only need to use the same number of nodes, message loss rate and run the experiments for 30 minutes.
To me SWIM is a new topic. So, after all. No new parameters are required. Sorry for the confusion.

File: src/cart/crt_swim.c
Comment (ashleypittman) [reply src/cart/crt_swim.c:327]: Then use strtol().

File: src/cart/utils/fault-inject-cart.yaml
Comment (dmiter) [reply src/cart/utils/fault-inject-cart.yaml:58]: This patch already contain all functionality for proposed experiments. This patch is for landing to master for future experiments and testing.

Comment (dmiter) [reply src/cart/utils/fault-inject-cart.yaml:58]: @berserk-fury could you approve it please if it works for you?

**Review by berserk-fury — APPROVED** [review]

**Comment by dmiter** [comment]
OSAOnlineDrain - DAOS-6061
DaosCoreTestDfs - DAOS-6090
Functional on Leap 15 - DAOS-6111

**Comment by dmiter** [comment]
@frostedcmos could you inspect it please?

File: src/cart/crt_swim.c
Comment (frostedcmos) [reply src/cart/crt_swim.c:327]: I agree with Ashley here; we should change this to strtol() when parsing arg/int32 and return code as is from config file instead of negating it

File: src/cart/crt_swim.c
Comment (frostedcmos) [inline src/cart/crt_swim.c:275]: I am not sure how you determine one way rpc via opcode like this; i couldnt find place on code where we change opcode based on whether it is one way rpc or not.

Should instead check for: 
rpc_priv->crp_opc_info->coi_no_reply

File: src/cart/crt_swim.c
Comment (frostedcmos) [inline src/cart/crt_swim.c:36]: I am not sure I like this; it becomes hard to correlate what is in yaml file
instead it is best in my view to pick/reserve some range of swim fault ids, for example from 65000 to 66000

File: src/cart/crt_swim.c
Comment (frostedcmos) [inline src/cart/crt_swim.c:94]: This is fine for now, but in future we might want to add D_ALWAYS_PRINT or some such instead of hijacking 'error' print for these purposes. We've had situations already when end-users and our devs were confused by errors being emitted by swim that are not errors.

**Comment by frostedcmos** [comment]
Also commit message needs to be updated, as it specifies "CRT_SWIM_FAIL_ID" environment variable which is now passed as argument param instead

File: src/cart/crt_swim.c
Comment (dmiter) [reply src/cart/crt_swim.c:36]: Hmm. In this case we should have a single list of those ID's reservation. In which header file this should be defined?
Do we have a recommended logic to allocate ids?

File: src/cart/crt_swim.c
Comment (dmiter) [reply src/cart/crt_swim.c:42]: I add conditional SWIM specific. So, I cannot change D_ macro.

File: src/cart/crt_swim.c
Comment (dmiter) [reply src/cart/crt_swim.c:74]: ok.

File: src/cart/crt_swim.c
Comment (dmiter) [reply src/cart/crt_swim.c:94]: Agree. But this print should go into the log file (not just on screen). My assumption about word "print" is just screen output.

File: src/cart/crt_swim.c
Comment (dmiter) [reply src/cart/crt_swim.c:275]: Hmm. Good ctach! I will loot at this and replace if it will work properly.

File: src/cart/crt_swim.c
Comment (dmiter) [reply src/cart/crt_swim.c:327]: Can we do this in separate PR?

File: src/cart/crt_swim.c
Comment (frostedcmos) [reply src/cart/crt_swim.c:36]: There is really no good way of doing this now, so we probably want to start documenting list of fault injection ids/reservations inside of the sample fault injection yaml file that we use.

File: src/cart/crt_swim.c
Comment (frostedcmos) [reply src/cart/crt_swim.c:42]: What I mean here is why can't you do something like this instead?
crt_swim_should_fail && (crt_swim_fail_id == id) && D_SHOULD_FAIL(fa) ?

File: src/cart/crt_swim.c
Comment (frostedcmos) [reply src/cart/crt_swim.c:94]: Correct, it should go to log file as well. In cart tests we use DBG_PRINT() for example that also goes to the log file. We can pick the appropriate naming for this like D_LOG_ALWAYS instead maybe to make it clear.

File: src/cart/crt_swim.c
Comment (frostedcmos) [reply src/cart/crt_swim.c:327]: Sure, please file a task jira ticket then on this so that this does not get forgotten

File: src/cart/crt_swim.c
Comment (dmiter) [reply src/cart/crt_swim.c:42]: The D_SHOULD_FAIL(fa) is not just conditions. It also has a THEN block. So, I don't want to have additional overhead on every  check.

File: src/cart/crt_swim.c
Comment (dmiter) [reply src/cart/crt_swim.c:327]: DAOS-6242

File: src/cart/crt_swim.c
Comment (ashleypittman) [reply src/cart/crt_swim.c:42]: The code block in D_SHOULD_FAIL is only executed if fault injection is turned on, and does logging which is something that we do want to happen, and in a consistent way.

File: src/cart/crt_swim.c
Comment (ashleypittman) [reply src/cart/crt_swim.c:94]: I don't see why a regular D_ERROR is not sufficient here, but we also have CRIT and ALARM I believe.

**Comment by ashleypittman** [comment]
It's probably a good idea to rebase this as there have been fault injection changes landed in the last few days.

File: src/cart/crt_swim.c
Comment (dmiter) [reply src/cart/crt_swim.c:74]: Hmm. It's properly aligned in final version (in my editor at least). You see this issue after adding "+ " for review.

**Comment by berserk-fury** [comment]
If I recall correctly. The new default build type is release ( `BUILD_TYPE=release`) where Fault Injection is disabled. Is there documentation about how to enable/use FI?
TTBOMK in order to use FI `BUILD_TYPE=dev` must be set and the vars `D_FI_CONFIG` and `CRT_SWIM_FAIL_ID` must be set as well. It would be nice to have such documentation.

File: src/cart/crt_swim.c
Comment (ashleypittman) [reply src/cart/crt_swim.c:330]: https://github.com/daos-stack/daos/pull/4041

File: src/cart/crt_swim.c
Comment (frostedcmos) [reply src/cart/crt_swim.c:94]: D_ERROR is sufficient, but it is misleading when things show up as 'error' in the log, while we just want to print them always and simply using D_ERROR to force that behavior for non-error messages

**Review by frostedcmos — APPROVED** [review]

**Comment by dmiter** [comment]
Can you inspect please?

**Review by berserk-fury — APPROVED** [review]
Looks good to me, however I am no expert on this regard so someone else should review it.

File: src/cart/crt_swim.c
Comment (jolivier23) [reply src/cart/crt_swim.c:94]: Even ERR isn't guaranteed to print.   No reason a user couldn't do log_mask: CRIT.  If you actually want it to never be skipped, you will need to add a new type.   We don't have any bits but we don't use ALERT or NOTE anywhere....we could potentially move the bits down, hijack bit 7 and make it so the user can only set EMRG or lower as the mask.   There is still the issue of a user setting D_SUBSYS to disable messages though...

File: src/cart/crt_swim.c
Comment (jolivier23) [reply src/cart/crt_swim.c:663]: I tend to agree on not adding the ***.   There should be a facility or some string to search instead.   If you added "IMPORTANT" or "ALWAYS" type, that would suffice as well.

**Pull request merged**
